### ISSUE TITLE
Support DNS changing IP address

### ISSUE THREAD

**Issue by cep21** [cactus/go-statsd-client#40]
We use DNS to resolve a statsd endpoint IP.  If the IP address behind the DNS changes, go-statsd-client will never use the new IP.  This means metrics will fail until the application restarts.  Ideally if the IP address behind a DNS name changes, the sender will send stats to the new IP.

It seems to start at this line:
```
	ra, err := net.ResolveUDPAddr("udp", addr)
```

Once `ra` is resolved, it is never re-resolved.  There are a few options for this.

1. Add a function to `SimpleSender` that allows me to update `ra`.  I could then call this if I detect a change.  This won't fix the problem for future people, but will let me work around it.
2. Change `SimpleSender` to execute `net.ResolveUDPAddr("udp", addr)` each time on Send: this could cause a performance problem.  Optionally, we could only update `ra` every X units of time.
3. Modify `BufferedSender` to take a `Sender` as input.  I could then pass my own sender, instead of using the provided one.

Also, any way we change this, we'll need to extend `BufferedSender` to call this function on the `SimpleSender`, or allow `BufferedSender` to return the wrapped `Sender`, and I could call it myself on the `BufferedSender` object.

**Comment by manaspaldhe12** [comment]
Smira's Go-statsd implements the feature to allows configuring`ReconnectInterval` - https://github.com/smira/go-statsd/blob/master/options.go#L73
It will be useful to have a similar configuration option in this library.

**Comment by cactus** [comment]
Hmm. You could currently work around this by defining your own SimpleSender, wrapping that in a BufferedSender, and passing it to `NewClientWithSender`. A lot of work for someone to get there though. 

I think adding something to the library to reconnect on an Interval, as @manaspaldhe12 referenced, would probably be the sanest option. For people who want something else, they would have a means to get there on their own. That also allows the library use to decide on their own a performance vs liveness tradeoff.

Perhaps the SimpleSender could retain the original address (add it to the struct), and add a method for "Reconnect"ing.

I'll try a few things when I get some time, and see how I feel about the end results.

**Comment by cactus** [comment]
I have some WIP code in https://github.com/cactus/go-statsd-client/tree/reconnecting-senders if  you are interested in checking it out.

``` go
import (
    "log"

    "github.com/cactus/go-statsd-client/statsd"
)

func main() {
    // A client that re-resolves the hostname ever 30 seconds. Useful if the address
    // of a hostname changes frequently. Note that this type of client has some
    // additional locking overhead for safety.
    // If you don't specifically need this functionality, use NewClient instead.
    client, err := statsd.NewResolvingClient("host.example.com:8125", "test-client", 30*time.Second)

    // A buffered resolving client, which sends multiple stats in one packet, as well
    // as re-resolves the hostname every 30 seconds.
    // If you don't specifically need this functionality, use NewBufferedClient instead.
    // client, err := statsd.NewBufferedResolvingClient("127.0.0.1:8125", "test-client", 30*time.Second, 300*time.Millisecond, 0)

    // handle any errors
    if err != nil {
        log.Fatal(err)
    }
    // make sure to clean up
    defer client.Close()

    // Send a stat
    client.Inc("stat1", 42, 1.0)
}
```

**Comment by cep21** [comment]
Looks good, just a few comments.  Mind making a pull request?

**Comment by cactus** [comment]
@cep21 sure thing. pr opened for feedback. Sorry it took a while, been busy at work.

**Comment by cactus** [comment]
Sorry this has taken so long. I'll give the code I have open in the PR another poke this weekend.

**Comment by cactus** [comment]
closed by 47b6058c80f53b9d6f8af839bf82de17cf80e1f2
